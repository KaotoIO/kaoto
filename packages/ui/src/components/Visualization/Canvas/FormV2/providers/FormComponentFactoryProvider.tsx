import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
