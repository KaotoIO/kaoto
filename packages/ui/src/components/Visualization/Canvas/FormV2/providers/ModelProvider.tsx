import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
