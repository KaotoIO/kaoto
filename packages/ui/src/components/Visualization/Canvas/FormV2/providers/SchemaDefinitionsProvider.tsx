import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
