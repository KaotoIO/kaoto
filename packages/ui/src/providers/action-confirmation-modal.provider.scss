.pf-v5-c-modal-box {
  &__title-text {
    white-space: normal;
  }
}
