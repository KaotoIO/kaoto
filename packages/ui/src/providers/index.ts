export * from './catalog-schema-loader.provider';
export * from './entities.provider';
