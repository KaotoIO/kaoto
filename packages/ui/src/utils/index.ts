export * from './camel-case-to-space';
export * from './camel-uri-helper';
export * from './catalog-schema-loader';
export * from './create-camel-properties-sorter';
export * from './event-notifier';
export * from './get-array-property';
export * from './get-custom-schema-from-kamelet';
export * from './get-field-groups';
export * from './get-required-properties-schema';
export * from './get-serialized-model';
export * from './get-user-updated-properties-schema';
export * from './get-value';
export * from './get-viznodes-from-graph';
export * from './init-visible-flows';
export * from './is-datamapper';
export * from './is-defined';
export * from './is-enum-type';
export * from './is-to-processor';
export * from './is-xslt-component';
export * from './join-path';
export * from './node-icon-resolver';
export * from './pipe-custom-schema';
export * from './promise-timeout';
export * from './resolve-ref-if-needed';
export * from './set-value';
export * from './update-kamelet-from-custom-schema';
export * from './weight-schema-properties';
