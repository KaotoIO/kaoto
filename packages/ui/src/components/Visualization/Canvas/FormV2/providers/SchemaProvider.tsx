import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
