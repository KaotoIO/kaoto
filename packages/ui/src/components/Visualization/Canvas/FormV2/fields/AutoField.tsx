import { FunctionComponent, useContext } from 'react';
import { CanvasFormTabsContext } from '../../../../../providers/canvas-form-tabs.provider';
import { isDefined } from '../../../../../utils';
import { useFieldValue } from '../hooks/field-value';
import { FormComponentFactoryContext } from '../providers/FormComponentFactoryProvider';
import { SchemaContext } from '../providers/SchemaProvider';
import { FieldProps } from '../typings';

export const AutoField: FunctionComponent<FieldProps> = ({ propName, required, onRemove }) => {
  const { selectedTab } = useContext(CanvasFormTabsContext);
  const { schema } = useContext(SchemaContext);
  const { value } = useFieldValue<object>(propName);
  const formComponentFactory = useContext(FormComponentFactoryContext);

  if (Object.keys(schema).length === 0) {
    throw new Error(`AutoField: schema is not defined for ${propName}`);
  } else if (!isDefined(formComponentFactory)) {
    throw new Error(`AutoField: formComponentFactory is not defined for ${propName}`);
  }

  const isFieldDefined = isDefined(value) && Object.keys(value).length > 0;
  const isComplexFieldType =
    schema.type === 'object' || schema.type === 'array' || 'oneOf' in schema || 'anyOf' in schema;
  // If required is not defined, it is considered as required
  const isFieldRequired = !isDefined(required) || (isDefined(required) && required);
  const isPropertiesField = schema.type === 'object' && Object.keys(schema?.properties ?? {}).length === 0;
  const shouldLoadField = isFieldRequired || (isDefined(value) && isComplexFieldType && !isPropertiesField);

  if ((selectedTab === 'Required' && !shouldLoadField) || (selectedTab === 'Modified' && !isFieldDefined)) {
    return null;
  }

  const FieldComponent = formComponentFactory(schema);

  return <FieldComponent propName={propName} required={required} onRemove={onRemove} />;
};
