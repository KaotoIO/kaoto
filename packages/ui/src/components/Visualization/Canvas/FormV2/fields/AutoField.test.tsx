import { act, render, screen } from '@testing-library/react';
import { inspect } from 'node:util';
import { CanvasFormTabsContext, CanvasFormTabsContextResult } from '../../../../../providers';
import { ROOT_PATH } from '../../../../../utils';
import { FormComponentFactoryContext, FormComponentFactoryProvider } from '../providers/FormComponentFactoryProvider';
import { ModelContextProvider } from '../providers/ModelProvider';
import { SchemaProvider } from '../providers/SchemaProvider';
import { AutoField } from './AutoField';

describe('AutoField', () => {
  let requiredValue: CanvasFormTabsContextResult;
  let modifiedValue: CanvasFormTabsContextResult;

  beforeEach(() => {
    requiredValue = { selectedTab: 'Required', onTabChange: jest.fn() };
    modifiedValue = { selectedTab: 'Modified', onTabChange: jest.fn() };
  });

  it('should throw an error if schema is not defined', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    expect(() => render(<AutoField propName={ROOT_PATH} />)).toThrow(
      `AutoField: schema is not defined for ${ROOT_PATH}`,
    );
  });

  it('should throw an error if formComponentFactory is not defined', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    expect(() =>
      render(
        <SchemaProvider schema={{ type: 'string' }}>
          <AutoField propName={ROOT_PATH} />{' '}
        </SchemaProvider>,
      ),
    ).toThrow(`AutoField: formComponentFactory is not defined for ${ROOT_PATH}`);
  });

  it('it should not render when in `Required` mode but no required properties', () => {
    const wrapper = render(
      <CanvasFormTabsContext.Provider value={requiredValue}>
        <FormComponentFactoryProvider>
          <SchemaProvider schema={{ type: 'object', properties: { name: { type: 'string' } } }}>
            <AutoField propName={ROOT_PATH} />
          </SchemaProvider>
        </FormComponentFactoryProvider>
      </CanvasFormTabsContext.Provider>,
    );

    expect(wrapper.container).toMatchSnapshot();
  });

  describe('Required mode', () => {
    it('it should render required fields', () => {
      const wrapper = render(
        <CanvasFormTabsContext.Provider value={requiredValue}>
          <FormComponentFactoryProvider>
            <SchemaProvider
              schema={{
                type: 'object',
                properties: { name: { type: 'string' }, lastName: { type: 'string' } },
                required: ['name'],
              }}
            >
              <AutoField propName={ROOT_PATH} />
            </SchemaProvider>
          </FormComponentFactoryProvider>
        </CanvasFormTabsContext.Provider>,
      );

      const inputFields = wrapper.queryAllByRole('textbox');

      expect(inputFields).toHaveLength(1);
    });

    it('it should render nested required fields from non-required but defined object parent', () => {
      const wrapper = render(
        <CanvasFormTabsContext.Provider value={requiredValue}>
          <FormComponentFactoryProvider>
            <SchemaProvider
              schema={{
                type: 'object',
                properties: { options: { type: 'object', properties: { key: { type: 'string' } }, required: ['key'] } },
              }}
            >
              <ModelContextProvider model={{ options: { key: 'MainKey' } }} onPropertyChange={jest.fn()}>
                <AutoField propName={ROOT_PATH} />
              </ModelContextProvider>
            </SchemaProvider>
          </FormComponentFactoryProvider>
        </CanvasFormTabsContext.Provider>,
      );

      const inputFields = wrapper.queryAllByRole('textbox');

      expect(inputFields).toHaveLength(1);
    });

    it('it should render nested required fields from non-required but defined array parent', async () => {
      await act(async () => {
        render(
          <CanvasFormTabsContext.Provider value={requiredValue}>
            <FormComponentFactoryProvider>
              <SchemaProvider
                schema={{
                  type: 'object',
                  properties: {
                    keys: {
                      type: 'array',
                      items: { type: 'object', properties: { name: { type: 'string' } }, required: ['name'] },
                    },
                  },
                }}
              >
                <ModelContextProvider model={{ keys: [{ name: 'MainKey' }] }} onPropertyChange={jest.fn()}>
                  <AutoField propName={ROOT_PATH} />
                </ModelContextProvider>
              </SchemaProvider>
            </FormComponentFactoryProvider>
          </CanvasFormTabsContext.Provider>,
        );
      });

      const inputFields = await screen.findAllByRole('textbox');
      expect(inputFields).toHaveLength(1);
    });

    it('it should render nested required fields from oneOf', () => {
      const wrapper = render(
        <CanvasFormTabsContext.Provider value={requiredValue}>
          <FormComponentFactoryProvider>
            <SchemaProvider
              schema={{
                oneOf: [
                  {
                    type: 'object',
                    properties: {
                      options: { type: 'object', properties: { key: { type: 'string' } }, required: ['key'] },
                    },
                  },
                ],
              }}
            >
              <ModelContextProvider model={{ options: { key: 'MainKey' } }} onPropertyChange={jest.fn()}>
                <AutoField propName={ROOT_PATH} />
              </ModelContextProvider>
            </SchemaProvider>
          </FormComponentFactoryProvider>
        </CanvasFormTabsContext.Provider>,
      );

      const inputFields = wrapper.queryAllByRole('textbox');

      expect(inputFields).toHaveLength(1);
    });

    it('it should render nested required fields from anyOf', () => {
      const wrapper = render(
        <CanvasFormTabsContext.Provider value={requiredValue}>
          <FormComponentFactoryProvider>
            <SchemaProvider
              schema={{
                type: 'object',
                anyOf: [
                  {
                    type: 'object',
                    properties: {
                      options: { type: 'object', properties: { lastname: { type: 'string' } }, required: ['lastname'] },
                    },
                  },
                ],
                properties: { name: { type: 'string' } },
              }}
            >
              <ModelContextProvider model={{ options: { lastname: 'Smith' } }} onPropertyChange={jest.fn()}>
                <AutoField propName={ROOT_PATH} />
              </ModelContextProvider>
            </SchemaProvider>
          </FormComponentFactoryProvider>
        </CanvasFormTabsContext.Provider>,
      );

      const inputFields = wrapper.queryAllByRole('textbox');
      expect(inputFields).toHaveLength(1);
      expect(inputFields[0]).toHaveValue('Smith');
    });
  });

  it('it should not render when in `Modified` mode with no modified properties', () => {
    const wrapper = render(
      <CanvasFormTabsContext.Provider value={modifiedValue}>
        <FormComponentFactoryProvider>
          <SchemaProvider schema={{ type: 'object', properties: { name: { type: 'string' } } }}>
            <ModelContextProvider model={{}} onPropertyChange={jest.fn()}>
              <AutoField propName={ROOT_PATH} />
            </ModelContextProvider>
          </SchemaProvider>
        </FormComponentFactoryProvider>
      </CanvasFormTabsContext.Provider>,
    );

    expect(wrapper.container).toMatchSnapshot();
  });

  it('it should render when in `Modified` mode, properties with value', () => {
    const wrapper = render(
      <CanvasFormTabsContext.Provider value={modifiedValue}>
        <FormComponentFactoryProvider>
          <SchemaProvider
            schema={{
              type: 'object',
              properties: {
                name: { type: 'string' },
                lastName: { type: 'string' },
              },
            }}
          >
            <ModelContextProvider model={{ name: 'test' }} onPropertyChange={jest.fn()}>
              <AutoField propName={ROOT_PATH} />
            </ModelContextProvider>
          </SchemaProvider>
        </FormComponentFactoryProvider>
      </CanvasFormTabsContext.Provider>,
    );

    const inputFields = wrapper.queryAllByRole('textbox');

    expect(inputFields).toHaveLength(1);
  });

  it('should get the component to render from the fromComponentFactory callback', () => {
    const factorySpy = jest.fn().mockReturnValue(() => <input aria-label="test" name="name" />);

    render(
      <FormComponentFactoryContext.Provider value={factorySpy}>
        <SchemaProvider schema={{ type: 'string' }}>
          <ModelContextProvider model="test" onPropertyChange={jest.fn()}>
            <AutoField propName={ROOT_PATH} />
          </ModelContextProvider>
        </SchemaProvider>
      </FormComponentFactoryContext.Provider>,
    );

    expect(factorySpy).toHaveBeenCalledWith({ type: 'string' });
  });

  it('should provide properties to the rendered field', () => {
    const factorySpy = jest.fn().mockReturnValue((props: Record<string, string>) => <pre>{inspect(props)}</pre>);

    const wrapper = render(
      <FormComponentFactoryContext.Provider value={factorySpy}>
        <SchemaProvider schema={{ type: 'string' }}>
          <ModelContextProvider model="test" onPropertyChange={jest.fn()}>
            <AutoField propName={ROOT_PATH} />
          </ModelContextProvider>
        </SchemaProvider>
      </FormComponentFactoryContext.Provider>,
    );

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
