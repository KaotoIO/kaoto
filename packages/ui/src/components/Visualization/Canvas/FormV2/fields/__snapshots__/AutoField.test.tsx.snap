// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoField Required mode it should not render object with undefined properties 1`] = `<div />`;

exports[`AutoField it should not render object with undefined properties in \`Modified\` mode with no modified properties 1`] = `<div />`;

exports[`AutoField it should not render when in \`Modified\` mode with no modified properties 1`] = `<div />`;

exports[`AutoField it should not render when in \`Required\` mode but no required properties 1`] = `<div />`;

exports[`AutoField should provide properties to the rendered field 1`] = `
<DocumentFragment>
  <pre>
    { propName: '#', required: undefined, onRemove: undefined }
  </pre>
</DocumentFragment>
`;
